### ISSUE TITLE
Rudimentary video support

### ISSUE THREAD

**Issue by tomaszg7** [BestImageViewer/geeqie#510]
It's just an idea, but probably some basic solution could be possible. The issue is that when using Geeqie to browse photos from digital camera, you're missing on the videos which were also taken. My workaround was to add several extensions (.mp4;.3gp;.mvi;.mov;.mts) to the file filter and created an editor desktop file which just runs mplayer/mpv. It is of course far from the perfect solution, but it works for me.

I have several ideas how to make it better:

1. Create new file class "video" and teach Geeqie to display it as a some kind of "play" icon, which upon clicking would start mplayer (or another player of user's choice)
2. Add some thumbnail generation support.
3. Add possibility to play the video directly in Geeqie window.

I don't know if 2 and 3 are feasible nor even if it would be a good idea (taking into account all extra dependencies and complications). However at least 1. should be quite straightforward.

**Comment by caclark** [comment]
I also don't think Geeqie should get involved with video playback. But as you say it can be easily done with a desktop file. I find that solution sufficient.

But I do think that a thumbnail display would be very useful. Maybe GnomeDesktopThumbnailFactory could be used.

**Comment by tomaszg7** [comment]
Yeah, but that means pulling a lot of Gnome dependencies... As far as I know it depends on GStreamer and it might be possible to use it directly (or ffmpeg, but from what I hear they have a terrible API). Or use something like https://github.com/dirkvdb/ffmpegthumbnailer

However I think that it might be worthwhile to make the use of videos easier for entry-level users, some build-in configurable support might be nice - even if it only resolves to replacing "broken thumbnail" with some generic video icon and adding UI support for starting playback (without bothering user with creating desktop file).

BTW. It would be nice if "metadata" class files didn't show that "broken thumbnail" but something more meaningful, some kind of "data" icon.

**Comment by caclark** [comment]
> BTW. It would be nice if "metadata" class files didn't show that "broken thumbnail" but something more meaningful, some kind of "data" icon.


That's already noted at line 241 of pixbuf_util.c
It doesn't look too difficult. Just needs some suitable .pngs for the format classes.

**Comment by tomaszg7** [comment]
Agreed, it is not too difficult, here's the patch together with some PNGs (as well as SVG variants):

[format_icons_patch.tar.gz](https://github.com/BestImageViewer/geeqie/files/1297603/format_icons_patch.tar.gz)

I added new "video" class while doing it, and included value of class in FileData struct. I also added a function for determining the class, right now it is not very useful, but it seems useable. 

(I think that metadata.c line 398 should be updated as well, but didn't want to put too many things into one patch)

The new icons aren't very good, I'm no graphic designer, but seem sufficient and consistent with UI.
Maybe it would make sense to migrate to SVGs instead. It's also probably not too difficult.

**Comment by caclark** [comment]
Icons implemented in commit d33e0f6aaf at geeqie.org

Sorry, Tomasz. I forgot to list you as the author.

**Comment by tomaszg7** [comment]
Ok, I think I will survive this ;)

Here are 3 small follow-up patches:

1. Add video files extension to default file filter (disabled by default). I was considering adding them as one entry, but that might be not convenient. Also I didn't make the list very exhaustive. I tried to include the most common and ones that can reasonably be expected to be produced by digital cameras or phones.

2. Use recorded format_class in fd in metadata.c

3. Do not try to generate/read thumb for files which are not images. 

[format-icons_followup_patches.tar.gz](https://github.com/BestImageViewer/geeqie/files/1299980/format-icons_followup_patches.tar.gz)

**Comment by caclark** [comment]
Above implemented in commit 9be75e34311 at geeqie.org

**Comment by tomaszg7** [comment]
Do you think it is worthwhile to look at replacing pngs with svgs? One problem is that I don't see a possibility to easily inline them, they would have to be installed as separate files.

**Comment by caclark** [comment]
At the moment I'm trying to finish a user-defined toolbar facility. This will give the user the option to install a wider range of toolbar items, especially links to user-created desktop (editor) files.
The problem is I need an easy way to create more icons - simple ones just consisting of two or three letters would be sufficient, rather than intricate graphics.
I think Inkscape would be an easier way to do this, rather than Gimp.
In other words, yes, svgs rather than pngs would be good.

**Comment by tomaszg7** [comment]
I did a bit more work on the video support issue. The patch is not final yet, but it should work (at least in "normal" viewing mode). It adds an option in preferences->behaviour enabling starting video playback by left clicking on image/video thumbnail with a selected plugin/desktop file. Right now I use a simple desktop file starting mpv. 

[geeqie-video-play.patch.gz](https://github.com/BestImageViewer/geeqie/files/1377681/geeqie-video-play.patch.gz)

This patch shouldn't conflict with mouse back button support patch from issue #512  (even though they touch the same function in layout_image.c.

**Comment by caclark** [comment]
I get a segfault in add_video_menu(). 

`if (!option) return;`
solves the problem, but it might be that the code doesn't check for the initial case when geeqierc.xml doesn't have any entries for this?


The patch makes things easier to see the few clips I have - easier than mousing through the menus to get the correct desktop file.

**Comment by tomaszg7** [comment]
Hmm, I thought I checked for that (for sure I didn't add new entries by hand). Must be that the bug crept in at some later point ;) Seems the segfault comes from g_hash_table_lookup() but according to GTK docs (if I understand the tag "nullable" correctly) it shouldn't happen. In any case, the fix is easy. I'll post updated patch when I'll correct other stuff + add some docs.

One question is what should take precendence if both navigation by left click and play by left click are set. Right now, play takes precendence, as you can also navigate by other methods (middle click, scroll, keyboard).

I'm also affraid of a potential problem when editors change between the call of add_video_menu() and video_menu_cb(). It might be partially circumvented by passing the "eds" GList to the function as well. 

BTW. You didn't comment on my other patch with support of mouse back buttons. Do you think it is a good idea?

**Comment by tomaszg7** [comment]
Ok, here's the updated patch. 

[geeqie-video-play.3.patch.gz](https://github.com/BestImageViewer/geeqie/files/1380184/geeqie-video-play.3.patch.gz)

There was also a bit missing to add %video group to filefilter:

[geeqie-format-class-group.patch.gz](https://github.com/BestImageViewer/geeqie/files/1380227/geeqie-format-class-group.patch.gz)

**Comment by caclark** [comment]
>  what should take precendence

As these are mutually exclusive operations, perhaps radio buttons would be preferable. But that means a 3-state radio button. 

>  potential problem when editors change 

This isn't safety critical software- covering every possibility isn't necessary. It doesn't crash, that's the most important aspect.

>  comment on my other patch 

Ah, sorry. I was going to try to find a general solution, but got distracted elsewhere. (I also had to wait for a gaming mouse from ebay. Too many buttons and wheels - and even a blue light! Far too complicated for me :-(.  )

**Comment by tomaszg7** [comment]
> As these are mutually exclusive operations, perhaps radio buttons would be preferable. But that means a 3-state radio button.

As I think more about it I get convinced that the current solution is optimal. In a sense we have 3 situations possible:
1. only navigation
2. only playback
3. playback on videos, navigation on other files

> Ah, sorry. I was going to try to find a general solution, but got distracted elsewhere. (I also had to wait for a gaming mouse from ebay. Too many buttons and wheels - and even a blue light! Far too complicated for me :-(. )

I don't have that many buttons myself ;) Probably the best solution would be to make it possible to map also mouse events to actions. But I didn't look at the code yet. But that discussion probably belongs to a separate issue :)

**Comment by caclark** [comment]
Implemented in commit 414e01d4ab at geeqie.org

**Comment by tomaszg7** [comment]
It seems today is a day of fixing bugs in my code ;) After the one you tracked earlier, I found another one in video playback code. It segfaults when you click on the image area in an empty directory. Obviously image fd is not defined, thus checking image class fails.

[geeqie-video-play-cor.patch.gz](https://github.com/BestImageViewer/geeqie/files/1385906/geeqie-video-play-cor.patch.gz)

**Comment by caclark** [comment]
Implemented in commit a737324b at geeqie.org

**Comment by mzagrabe** [comment]
@tomaszg7 

Thanks for the video support! I just cloned the repo and compiled. I checked the boxes for the video file types, but there are no video players in the "play with" drop down.

Would you be willing to share your desktop file. I tried to create one, but it didn't show up in under "plugins" for geeqie, which is where I expected it.

My current desktop file is:

cat ~/.config/geeqie/applications/mpv.desktop
[Desktop Entry]
Name=mpv
GenericName=Video Player
Comment=Play a video file
Type=Application
Exec=mpv %F
Terminal=false

Thanks for any help!

-m

**Comment by tomaszg7** [comment]
It's nothing fancy, almost the same as yours. I don't know which of of those settings are important, probably the "Categories" one.
```
[Desktop Entry]
Version=1.0
Type=Application

Name=mpv
Icon=media-playback-start
Exec=mpv %f

Categories=X-Geeqie;
OnlyShowIn=X-Geeqie;
```
I just created it from default template. Maybe a sample video player should be included, but I don't know if we can expect any specific media player to be universally available.

**Comment by mzagrabe** [comment]
Works like a charm! Perhaps a mpv.desktop.example file in doc/ with a mention of:

cp doc/mpv.desktop.example ~/.config/geeqie/applications/mpv.desktop

in the README. I'm not sure if users would check the github issues tracker to create a desktop file.

Thanks again for your help and for writing free software.
